src/acm_pca_cert_generator/certgen.py

```diff
@@ -0,0 +1,417 @@
+        "--signing-algorithm",
+        choices=[
+            "SHA256WITHECDSA",
+            "SHA384WITHECDSA",
+            "SHA512WITHECDSA",
+            "SHA256WITHRSA",
+            "SHA384WITHRSA",
+            "SHA512WITHRSA",
+        ],
+        required=True,
+        env_var="CERTGEN_SIGNING_ALGORITHM",
+        help="The algorithm that ACM PCA will use to sign the certificate",
+    )
+    p.add(
+        "--validity-period",
+        required=True,
+        type=check_validity_period,
+        env_var="CERTGEN_VALIDITY_PERIOD",
+        help="How long the certificate is valid for, e.g. 1d, 1m, 1y for 1 "
+        "day, 1 month and 1 year respectively",
+    )
+    p.add("--log-level", default="DEBUG", env_var="CERTGEN_LOG_LEVEL")
+
+    return p.parse_args(args)
+
+
+def generate_private_key(key_type, key_bits):
+    """Generate a private SSL key.
+
+    Args:
+        key_type (str): The type of key to generate (RSA, DSA)
+        key_bits (int): The length of the key
+
+    Returns:
+        OpenSSL.Crypto.PKey: The OpenSSL private key
+
+    """
+    key = OpenSSL.crypto.PKey()
+    if key_type == "RSA":
+        openssl_key_type = OpenSSL.crypto.TYPE_RSA
+    elif key_type == "DSA":
+        openssl_key_type = OpenSSL.crypto.TYPE_DSA
+    else:
+        raise ValueError(
+            "Invalid value for key_type. Only 'RSA' and 'DSA' are supported."
+        )
+    key.generate_key(openssl_key_type, key_bits)
+    return key
+
+
+def generate_csr(pkey, digest, subject_details):
+    """Generate a certificate signing request.
+
+    Args:
+        pkey (OpenSSL.crypto.PKey): The key pair to sign the CSR with
+        digest (str): The name of the message digest to use for the signature
+        subject_details (dict): A dict containing all of the certificate subject's
+                                details. Must contain keys "C", ST", "L", "O", "OU",
+                                "CN", "emailAddress"
+
+    Return:
+        str: The PEM-encoded CSR
+
+    Raises:
+        KeyError: If subject_details doesn't contain all of the required keys
+
+    """
+    csr = OpenSSL.crypto.X509Req()
+    subject = csr.get_subject()
+
+    subject_name_parts = ["C", "ST", "L", "O", "OU", "CN", "emailAddress"]
+    for name_part in subject_name_parts:
+        setattr(subject, name_part, subject_details[name_part])
+
+    csr.set_pubkey(pkey)
+    csr.sign(pkey, digest)
+    return OpenSSL.crypto.dump_certificate_request(OpenSSL.crypto.FILETYPE_PEM, csr)
+
+
+def create_validity_dict(validity_period):
+    """Convert a validity period string into a dict for issue_certificate().
+
+    Args:
+        validity_period (str): How long the signed certificate should be valid for
+
+    Returns:
+        dict: A dict {"Value": number, "Type": "string" } representation of the
+              validity period
+
+    """
+    validity_length = int(validity_period[:-1])
+    validity_suffix = validity_period[-1:]
+    if validity_suffix == "d":
```

Comment:
there is no default case. Default should probably be to raise and exception. Currently the return could hit a `validity_unit` not set error

Reply:
I had a chat with @nicholasjgreen about this; at the moment this module is only expected to be called from the CLI, which does enough argument validation to ensure this can't happen.  Happy to have a wider discussion about this though, as I'm coming around to thinking that the function itself should validate all of its args and raise exceptions as it sees fit.

Reply:
if you want to lint this code when building you will probably have to add an exception for this. It should never reach a not set error, but changes to the validator function could accidentally allow it